### PR TITLE
Cleaned up MemoryPool

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
Renamed BufferPool to MemoryPool\<T\>
Made the pool generic
Cleaned up file locations, and some naming

@davidfowl, what do you think about this as the abstraction?

It would be good to make pipelines MemoryPool internal (and just have a factory on Pipe). I think it could be done if we add callbacks to MemoryPool<T> for when memory is allocated. The callbacks would give you access to OwnedMemory<T>, but I think it's all you need, i.e. Length and native pointer.

cc: @ahsonkhan, @pakrym,

### CODE REVIEW

File: src/System.Buffers.Primitives/System/Buffers/OwnedArray.cs [hunk 1]
```diff
@@ -19,7 +20,7 @@ public OwnedArray(int length)
 
         public OwnedArray(T[] array)
         {
-            if (array == null) BufferPrimitivesThrowHelper.ThrowArgumentNullException(nameof(array)); 
+            if (array == null) new ArgumentNullException(nameof(array));
```
Comment (ahsonkhan) [inline src/System.Buffers.Primitives/System/Buffers/OwnedArray.cs]: nit: use throw helper here like in other places

File: src/System.IO.Pipelines/System/IO/Pipelines/PipeOptions.cs [hunk 2]
```diff
@@ -31,6 +31,6 @@ public class PipeOptions
 
         public IScheduler ReaderScheduler { get; }
 
-        public BufferPool BufferPool { get; }
+        public MemoryPool BufferPool { get; }
```
Comment (pakrym) [inline src/System.IO.Pipelines/System/IO/Pipelines/PipeOptions.cs]: Rename property too

File: src/System.IO.Pipelines/System/IO/Pipelines/PipeOptions.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ namespace System.IO.Pipelines
     public class PipeOptions
     {
         public PipeOptions(
-            BufferPool bufferPool,
+            MemoryPool bufferPool,
```
Comment (pakrym) [inline src/System.IO.Pipelines/System/IO/Pipelines/PipeOptions.cs]: Rename argument

File: src/System.IO.Pipelines/System/Buffers/MemoryPool.cs [hunk 4]
```diff
@@ -78,6 +78,12 @@ public override OwnedMemory<byte> Rent(int size)
             return block;
         }
 
+        public override void Return(OwnedMemory<byte> buffer)
+        {
+            var block = buffer as MemoryPoolBlock;
+            if (block != null) Return(block);
```
Comment (pakrym) [inline src/System.IO.Pipelines/System/Buffers/MemoryPool.cs]: Throw?

Comment (pakrym) [reply src/System.IO.Pipelines/System/Buffers/MemoryPool.cs]: I don't think we need this public, only thing that returns is `OnZeroReferences` of `MemoryPoolBlock`

File: src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs [hunk 5]
```diff
@@ -3,23 +3,25 @@
...
-    public abstract class BufferPool : IDisposable
+    public abstract class MemoryPool<T> : IDisposable
     {
-        public static BufferPool Default => Internal.ManagedBufferPool.Shared;
+        public static MemoryPool<T> Default => Internal.ArrayMemoryPool<T>.Shared;
 
-        public abstract OwnedMemory<byte> Rent(int minimumBufferSize);
+        public abstract OwnedMemory<T> Rent(int minimumBufferSize);
+
+        public abstract void Return(OwnedMemory<T> buffer);
```
Comment (pakrym) [inline src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: Do we need `Return` at all if we have `Release` in `OwnedMemory<T>`?

Comment (davidfowl) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: Yea, this was a point of contention before.

File: src/System.Buffers.Primitives/System/Buffers/OwnedArray.cs (hunk 1 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/OwnedArray.cs]: will do

File: src/System.IO.Pipelines/System/IO/Pipelines/PipeOptions.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines/System/IO/Pipelines/PipeOptions.cs]: will do

File: src/System.IO.Pipelines/System/IO/Pipelines/PipeOptions.cs (hunk 3 above)
Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines/System/IO/Pipelines/PipeOptions.cs]: will do

File: src/System.IO.Pipelines/System/Buffers/MemoryPool.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply src/System.IO.Pipelines/System/Buffers/MemoryPool.cs]: Yeah, I was thinking about it. I will change to throw.

File: src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs (hunk 5 above)
Comment (KrzysztofCwalina) [reply src/System.Buffers.Primitives/System/Buffers/Pooling/MemoryPool.cs]: I added the method to allow pooling of the OwnedMemory instances.

**Comment by KrzysztofCwalina** [comment]
Yeah, let's discuss. I added it to allow pooling OwnedMemory instances and for parity with other pools (e.g. ArrayPool). OwnedMemory instances are free to store the pool in them and call Return(this) when they are disposed. I.e. I don't see the negative of adding this method. Do you?

**Comment by pakrym** [comment]
It's a mixed message from abstraction, do you as consumer have to `Return` or `Release` is enough?

**Comment by benaadams** [comment]
> Yeah, let's discuss. I added it to allow pooling OwnedMemory instances and for parity with other pools (e.g. ArrayPool)

`ArrayPool<T>` the management is external and manual (as the array doesn't sidecar the pool); so you **can't** use a dispose pattern.

Should either use `Dispose` allowing a nice scoping with `using`, manual `Dispose` when longer lived; or manual `Return` but not both. 

Dispose is probably a better pattern to go for

**Comment by KrzysztofCwalina** [comment]
OK, the arguments make sense. I will remove Return.

**Comment by KrzysztofCwalina** [comment]
@dotnet-bot test Innerloop OSX10.12 Debug Build and Test

**Pull request merged**